utils/zerovec/src/zerovec/mod.rs

```diff
@@ -919,6 +919,87 @@ impl<T: AsULE> FromIterator<T> for ZeroVec<'_, T> {
     }
 }
 
+/// Convenience wrapper for [`ZeroSlice::from_ule_slice`]. The value will be created at compile-time.
+///
+/// # Arguments
+///
+/// * `$aligned` - The type of an element in its canonical, aligned form, e.g., `char`.
+/// * `$array_fn` - A const function that converts an array of `$aligned` elements into an array
+///                 of their unaligned equivalents, e.g.,
+///                 `const fn from_array<const N: usize>(arr: [char; N]) -> [<char as AsULE>::ULE; N]`.
+/// * `$x` - The elements that the `ZeroSlice` will hold.
+///
+/// # Examples
+///
+/// Using array-conversion functions provided by this crate:
+///
+/// ```
+/// use zerovec::{ZeroSlice, zeroslice, ule::AsULE};
+///
+/// const SIGNATURE: &ZeroSlice<char> = zeroslice![char; <char as AsULE>::ULE::from_array; 'b', 'y', 'e', '✌'];
+/// const EMPTY: &ZeroSlice<u32> = zeroslice![];
+/// let empty: &ZeroSlice<u32> = zeroslice![];
+/// let nums = zeroslice![u32; <u32 as AsULE>::ULE::from_array; 1, 2, 3, 4, 5];
+/// assert_eq!(nums.last().unwrap(), 5);
+/// ```
+///
+/// Using a custom array-conversion function:
+///
+/// ```
+/// use zerovec::{ZeroSlice, zeroslice};
+///
+/// mod conversion {
+///     use zerovec::ule::RawBytesULE;
+///     pub(super) const fn i16_array_to_be_array<const N: usize>(arr: [i16; N]) -> [RawBytesULE<2>; N] {
+///         let mut result = [RawBytesULE([0; 2]); N];
+///         let mut i = 0;
+///         while i < N {
+///             result[i] = RawBytesULE(arr[i].to_be_bytes());
+///             i += 1;
+///         }
+///         result
+///     }
+/// }
+///
+/// const NUMBERS: &ZeroSlice<i16> = zeroslice![i16; conversion::i16_array_to_be_array; 1, -2, 3, -4, 5];
+/// ```
+#[macro_export]
+macro_rules! zeroslice {
+    () => (
+        $crate::ZeroSlice::new_empty()
+    );
+    ($aligned:ty; $array_fn:expr; $($x:expr),+ $(,)?) => (
```

Comment:
thought: add a case without `$array_fn` that uses `<$aligned as $crate::ule::AsULE>::ULE::from_array`.

Reply:
Please have a look at #3454

Reply:
is the type strictly necessary? We could probably make do with using `_` if the type is omitted and rely on context

Reply:
We need the unaligned type to declare the const slice, and the cleanest way to get that is to ask for the aligned type.

Reply:
To expand on that, the internal `const X` is necessary for non-const contexts:

```
let x: &ZeroSlice<char> = zeroslice![<char as AsULE>::ULE::from_array; 'a'];
println!("{:?}", x); // error: creates a temporary value which is freed while still in use ^
```

Reply:
Fortunately or unfortunately, I can avoid the need for the temporary by using `transmute` directly instead of calling the ZeroSlice constructor...

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2ea56560dc8e192e744faf2f1abe60bb

Reply:
Would like @Manishearth's thoughts on if there is some other way we can avoid the type. It's totally something the compiler _should_ be able to infer; I just don't know _how_ in this context

Reply:
I don't think that code is sound. You're extending the lifetime of the *type*, but the temporary array still gets dropped at the end of the expression. In fact if you run Miri you get
```
error: Undefined Behavior: pointer to alloc8209 was dereferenced after this allocation got freed
```

Reply:
Inline `const` can solve this in the future. However in the follow up we're using the type to look up the conversion function, and I'd rather specify the type than the function.